### PR TITLE
Ensure invalid tunnel agent connections get closed

### LINKED ISSUE #15911: `remoteConn.markInvalid` does not lead to `remoteConn` getting closed.
Expected behavior:
If the reverse tunnel server marks a connection as invalid, said connection gets closed or somehow signaled so that the agent knows that it should reconnect.

Current behavior:
The connection is not closed and the agent thinks that everything is fine, potentially rendering it unaccessible via said proxy.

Might be the cause of #13673.

### PR DESCRIPTION
**Pull request by rosstimothy**
Connections from reverse tunnel agents were being marked as invalid by the proxy under certain conditions but would ultimately never be closed. This could lead to scenarios where the agent thought things were fine but the proxy considered that agent unhealthy and not routable.

Pruning of invalid connections used to occur when a proxy tried to retrieve a connection for that tunnel. This also further muddied the point in time at which the proxy could close a connection as it never explicitly stopped tracking the connection and closed it at the same time.

To remedy this, connections are explicitly closed by the proxy and removed from the mapping to stop tracking immediately. In order to prevent a connection that is servicing an active connection from being closed the proxy now tracks which connections have sessions. Closing does not occur when there are any active sessions to prevent them from being force terminated.

When the proxy receives a heartbeat from an agent it now restores the connection to a valid state. In the event that too many heart beats have been missed for an agent, the proxy will now terminate the connection, again only if it is not serving any sessions.

Fixes #15911

### CODE REVIEW

File: lib/reversetunnel/conn.go [hunk 1]
```diff
@@ -154,13 +157,33 @@ func (c *remoteConn) ChannelConn(channel ssh.Channel) net.Conn {
...
+}
+
 func (c *remoteConn) markInvalid(err error) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
 	c.lastError = err
 	atomic.StoreInt32(&c.invalid, 1)
-	c.log.Debugf("Disconnecting connection to %v %v: %v.", c.clusterName, c.conn.RemoteAddr(), err)
+	c.log.Debugf("Unhealthy connection to %v %v: %v.", c.clusterName, c.conn.RemoteAddr(), err)
```
Comment (jdconti) [inline lib/reversetunnel/conn.go:178]: What are your thoughts on changing the log level here? I feel like this would be a useful bit of debugging and shouldn't be too chatty (and if it is, it's likely indicative of an issue).

File: lib/reversetunnel/localsite.go [hunk 2]
```diff
@@ -602,23 +671,55 @@ func (s *localSite) chanTransportConn(rconn *remoteConn, dreq *sshutils.DialReq)
 	if err != nil {
 		if markInvalid {
 			rconn.markInvalid(err)
+			// If not serving any connections close and remove this connection immediately.
+			// Otherwise, let the heartbeat handler detect this connection is down.
+			if rconn.activeSessions() == 0 {
+				s.removeRemoteConn(rconn)
+				return nil, trace.NewAggregate(trace.Wrap(err), rconn.Close())
+			}
```
Comment (fspmarshall) [inline lib/reversetunnel/localsite.go:679]: Is this removal actually necessary?  IIUC removal here means that we still might hit the case where a conn may continue to hb, but isn't routable.  I think we're going to have an easier time if removal is tightly coupled with closure.

**Review by fspmarshall — APPROVED** [review]
LGTM w/ some minor nits/tweaks.

File: lib/reversetunnel/localsite.go [hunk 3]
```diff
@@ -544,54 +573,94 @@ func (s *localSite) handleHeartbeat(rconn *remoteConn, ch ssh.Channel, reqC <-ch
...
-	conns := s.remoteConns[key]
+	// Check the remoteConns from newest to oldest for one
+	// that has heartbeated and is valid. If none are valid, try
+	// the newest ready but invalid connection.
+	var newestInvalidConn *remoteConn
 	for i := len(conns) - 1; i >= 0; i-- {
-		if conns[i].isReady() {
+		switch {
+		case !conns[i].isReady(): // skip remoteConn that haven't heartbeated yet
+			continue
```
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/reversetunnel/localsite.go:647]: Did some poking around... doesn't look like waiting for the first *heartbeat* is necessary, but it does seem like it is necessary to wait for the heartbeat *channel* to be created, just to make sure that the agent has set up all its request/channel handlers.  Probably best to leave the logic here as-is.

In theory, I think we could eek out a bit of an improvement in setup time by just getting rid of the first hb sent by the agent, and treating the hb channel open *as* the first hb.  Out of scope for this PR I think, but something worth remembering if we ever decide tunnel setup time is a problem in the future.

[the remark this answers is not in the document]
Comment (espadolini) [reply lib/reversetunnel/localsite.go:647]: It wouldn't be much of an improvement, channel open and first heartbeat are one after another in the same goroutine with no extra sleeps.

I agree that the future protocol should tie the availability of the node side for new connections to the lifetime of a `reversetunnel@goteleport.com` channel rather than to some weird heartbeat system (for a couple of protocol versions, then we can decide if that should be a global request that irreversibly marks that ssh connection as available for reverse tunneling, or if we just leave it as is and figure out some TLS based way of doing it).

[the remark this answers is not in the document]
Comment (espadolini) [reply lib/reversetunnel/localsite.go:647]: On said hypothetical new reverse tunnel protocol we should just remove the entire heartbeat mechanism. The only value that it gives us is that an agent that's stopping is going to stop heartbeating and it's going to reject new connection attempts, so _maybe_ there's a point to treating a connection that's not heartbeating as a "lesser" one; however, the only true indicators that a connection is not actually usable for new connections anymore is a ssh error (potentially caused by closing the underlying connection due to idle timeouts?) or a reject on a `teleport-transport` channel open.

File: lib/reversetunnel/conn.go [hunk 4]
```diff
@@ -119,23 +122,24 @@ func (c *remoteConn) String() string {
 
 func (c *remoteConn) Close() error {
 	// If the connection has already been closed, return right away.
-	if !atomic.CompareAndSwapInt32(&c.closed, 0, 1) {
+	if !c.closed.CompareAndSwap(false, true) {
```
Comment (espadolini) [inline lib/reversetunnel/conn.go:125]:
```suggestion
	if c.closed.Swap(true) {
```

**Review by espadolini — APPROVED** [review]
